Fix regression generating configuration

BraisGabin opened the pull request:
This is not my nicest job for sure. But now it works and we have a test to ensure that it doesn't break again.

To fix this properly we need to do a huge refactor on the core.

fix #4601

BraisGabin commented on detekt-tooling/api/detekt-tooling.api:
```diff
@@ -11,12 +11,12 @@ public abstract interface class io/github/detekt/tooling/api/DefaultConfiguratio
-	public final fun load (Lio/github/detekt/tooling/api/spec/ProcessingSpec;Ljava/lang/ClassLoader;)Lio/github/detekt/tooling/api/DefaultConfigurationProvider;
-	public static synthetic fun load$default (Lio/github/detekt/tooling/api/DefaultConfigurationProvider$Companion;Lio/github/detekt/tooling/api/spec/ProcessingSpec;Ljava/lang/ClassLoader;ILjava/lang/Object;)Lio/github/detekt/tooling/api/DefaultConfigurationProvider;
+	public final fun load (Lio/github/detekt/tooling/api/spec/ExtensionsSpec;Ljava/lang/ClassLoader;)Lio/github/detekt/tooling/api/DefaultConfigurationProvider;
+	public static synthetic fun load$default (Lio/github/detekt/tooling/api/DefaultConfigurationProvider$Companion;Lio/github/detekt/tooling/api/spec/ExtensionsSpec;Ljava/lang/ClassLoader;ILjava/lang/Object;)Lio/github/detekt/tooling/api/DefaultConfigurationProvider;
```
This functins were introduced in 1.20 so it's safe to change them.

cortinico commented:
Could you provide some tests + more context on what this change is doing (it's not immediate from reading the code).

BraisGabin commented:
https://github.com/detekt/detekt/pull/4646/commits/ddd6adeb44808d58d93288e96d72ccca5aab6411#diff-ea3a27e29a40bcb62f6014e51bfdf793236b1ed156b98170570086a1cff78415 this is the unit test that I added to reproduce the issue. The test wasn't testing the real state. The real state is that the file doesn't exist and that all works. That case was failing before and know it works.

All the fix is in this commit: https://github.com/detekt/detekt/pull/4646/commits/ddd6adeb44808d58d93288e96d72ccca5aab6411 the other commits are just refactors to make this change easier. I don't know how to add more tests because the tests are already there. 

---

About context. The issue was this line of code that I added in #4315: 

https://github.com/detekt/detekt/blob/481b3b24288c77e62f8c5742d3d5ab4e4206d13f/detekt-cli/src/main/kotlin/io/gitlab/arturbosch/detekt/cli/runners/ConfigExporter.kt#L16

The problem is that `arguments.createSpec` validates that the configuration is ready to execute detekt. One of those checks is: If there is a configuration file path it should exist. But when you are generating the configuration the file doesn't need to exist. For that reason I can't use `arguments.createSpec`. That's the reason I'm using the Spec DSL directly. (Yeap, we have a DSL for that, Artur did so many things :P)

cortinico commented:
Thanks for the explanation. IMHO this can be merged 👍
